libsolidity/interface/CompilerStack.cpp

```diff
@@ -504,19 +504,28 @@ bool CompilerStack::compile()
 	if (m_hasError)
 		BOOST_THROW_EXCEPTION(CompilerError() << errinfo_comment("Called compile with errors."));
 
-	// Only compile contracts individually which have been requested.
-	map<ContractDefinition const*, shared_ptr<Compiler const>> otherCompilers;
-	for (Source const* source: m_sourceOrder)
-		for (ASTPointer<ASTNode> const& node: source->ast->nodes())
-			if (auto contract = dynamic_cast<ContractDefinition const*>(node.get()))
-				if (isRequestedContract(*contract))
-				{
-					compileContract(*contract, otherCompilers);
-					if (m_generateIR || m_generateEwasm)
-						generateIR(*contract);
-					if (m_generateEwasm)
-						generateEwasm(*contract);
-				}
+	try
+	{
+		// Only compile contracts individually which have been requested.
+		map<ContractDefinition const*, shared_ptr<Compiler const>> otherCompilers;
+		for (Source const* source : m_sourceOrder)
+			for (ASTPointer<ASTNode> const& node : source->ast->nodes())
+				if (auto contract = dynamic_cast<ContractDefinition const*>(node.get()))
+					if (isRequestedContract(*contract))
+					{
+						compileContract(*contract, otherCompilers);
+						if (m_generateIR || m_generateEwasm)
+							generateIR(*contract);
+						if (m_generateEwasm)
+							generateEwasm(*contract);
+					}
+	}
+	catch (FatalError const&)
```

Comment:
Do we catch fatal errors elsewhere in the compiler stack? I think the policy is that components can use fatal errors internally, but should not leak them to the compiler stack.

Reply:
Without the `catch`, we would end up in
```
	/// This should not be leaked from compile().
	catch (FatalError const& _exception)
	{
		errors.append(formatError(
			. . . .
			"Uncaught fatal error: " + boost::diagnostic_information(_exception)
		));
	}
```

It seems to me that currently `CompilerStack::compileContract` is supposed to report warnings only. And in case of real errors, we just throw, e,g
```
BOOST_THROW_EXCEPTION(CompilerError() << errinfo_comment("Given source file not found."));
```

Maybe we should change the architecture and allow regular errors during compilation/optimization too.

Reply:
`compileContract` already catches errors - can you catch the appropriate error there? If you want to use an exception, then a new class would be in order, I think.